### PR TITLE
[CPU][sgl-kernel] biased_grouped_topk: fix correction_bias dtype to float32

### PR DESCRIPTION
**Pull request by chunyuan-w**
## Motivation

https://github.com/sgl-project/sglang/pull/7825 has fixed the dtype of `e_score_correction_bias` to float32 instead of bfloat16. We need to fix the support in the topk kernel.
UT has been updated accordingly.

### CODE REVIEW

**Comment by chunyuan-w** [comment]
The CI failure is not related to the change in the current PR. 
Other PRs are having the same failures, for example:
https://github.com/sgl-project/sglang/actions/runs/16427578447/job/46425726939?pr=7135

File: sgl-kernel/csrc/cpu/topk.cpp [hunk 1]
```diff
@@ -644,16 +651,23 @@ std::tuple<at::Tensor, at::Tensor> biased_grouped_topk_cpu(
...
-        TORCH_CHECK(false, "Unexpected num_experts: ", num_experts);
-    }
-  });
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      correction_bias.scalar_type(),
+      "biased_grouped_topk_kernel2",
+      [&] {
+        using bias_scalar_t = scalar_t;
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/topk.cpp]: `scalar_t` will be dtype from `correction_bias` in `AT_DISPATCH_FLOATING_TYPES_AND2`.

yet in `LAUNCH_BIASED_GROUPED_TOPK_KERNEL`, it has `gating_output.data_ptr<scalar_t>()`

isn't this going to be problematic?

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
To make the code change more decent we can do something like this:

re-write `AT_DISPATCH_REDUCED_FLOATING_TYPES` in `common.h`, say 


set `scalar_t` and `param_t` in the MACRO, the logic is similar to mixed_dtype_support in BatchNorm2d in pytorch, let's use the similar naming convention.


```
// dispatch: bfloat16, float16, int8_t, fp8_e4m3
#define CPU_DISPATCH_REDUCED_FLOATING_TYPES_EXT(TYPE1, TYPE2 ...)                                    \
  [&] {                                                                         \
    if (TYPE2 == at::kFloat) {
    switch (TYPE1) {                                                             \
      case at::ScalarType::BFloat16 : {                                         \
        using scalar_t = at::BFloat16;                                          \
        using param_t = at::Float;                                        \
        return __VA_ARGS__();                                                   \
      }                                                                         \
      case at::ScalarType::Half: {                                              \
        using scalar_t = at::Half;                                              \
        using param_t = at::Float;             \
        return __VA_ARGS__();                                                   \
      }                                                                         \
     )
      default:                                                                  \
        TORCH_CHECK(false, "Unsupported floating data type.\n");                \
    }                                                                           \
    else {
      TORCH_CHECK(TYPE1 == TYPE2);
      switch (TYPE1) {
        // set `scalar_t` and `param_t`

      }
    }
  }()

```

**Comment by mingfeima** [comment]
also you can put a help function in `vec.h`

```
// base type
template <typename scalar_t, typename std::enable_if /* allow only f32, f16, bf16*/>
inline std::tuple<Vectorized<float>, Vectorized<float>>  load_float_vec2(const scalar_t* __restrict__ data) {...}

// specification for `float`
inline std::tuple<Vectorized<float>, Vectorized<float>>  load_float_vec2<float>(const float* __restrict__ data) {...}

```

then the code will be more neat.

**Comment by jianan-gu** [comment]
> To make the code change more decent we can do something like this:
> 
> re-write `AT_DISPATCH_REDUCED_FLOATING_TYPES` in `common.h`, say
> 
> set `scalar_t` and `param_t` in the MACRO, the logic is similar to mixed_dtype_support in BatchNorm2d in pytorch, let's use the similar naming convention.
> 
> ```
> // dispatch: bfloat16, float16, int8_t, fp8_e4m3
> #define CPU_DISPATCH_REDUCED_FLOATING_TYPES_EXT(TYPE1, TYPE2 ...)                                    \
>   [&] {                                                                         \
>     if (TYPE2 == at::kFloat) {
>     switch (TYPE1) {                                                             \
>       case at::ScalarType::BFloat16 : {                                         \
>         using scalar_t = at::BFloat16;                                          \
>         using param_t = at::Float;                                        \
>         return __VA_ARGS__();                                                   \
>       }                                                                         \
>       case at::ScalarType::Half: {                                              \
>         using scalar_t = at::Half;                                              \
>         using param_t = at::Float;             \
>         return __VA_ARGS__();                                                   \
>       }                                                                         \
>      )
>       default:                                                                  \
>         TORCH_CHECK(false, "Unsupported floating data type.\n");                \
>     }                                                                           \
>     else {
>       TORCH_CHECK(TYPE1 == TYPE2);
>       switch (TYPE1) {
>         // set `scalar_t` and `param_t`
> 
>       }
>     }
>   }()
> ```

Got it, will refine the code accorddingly, Thanks!

File: sgl-kernel/csrc/cpu/topk.cpp (hunk 1 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/topk.cpp]: Have updated this parts with following suggestions and fixed this issue. Thanks.

File: sgl-kernel/csrc/cpu/topk.cpp [hunk 2]
```diff
@@ -252,29 +252,33 @@ void topk_softmax_kernel_impl(
...
   using fVec = at::vec::Vectorized<float>;
-  for (int d = 0; d < SIZE; d += bVec::size()) {
-    bVec bias_vec = bVec::loadu(bias + d);
-    fVec bias0, bias1;
-    std::tie(bias0, bias1) = at::vec::convert_to_float(bias_vec);
-
-    fVec x0 = fVec::loadu(scores + d) + bias0;
-    fVec x1 = fVec::loadu(scores + d + fVec::size()) + bias1;
+  using bVec = at::vec::Vectorized<param_t>;
+  auto vec_size = fVec::size() * 2 == bVec::size() ? bVec::size() : fVec::size() * 2;
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/topk.cpp]: isn't this always 32 on avx512 for `vec_size`? just use:

```
template <typename scalar_t, typename param_t, int SIZE>
inline void apply_bias(...) {
  using bVec = att::vec::Vectorized<scalar_t>;
}
```
will do all the work.

File: sgl-kernel/csrc/cpu/vec.h [hunk 3]
```diff
@@ -16,6 +16,28 @@ inline Vectorized<scalar_t> convert_from_float_ext(const Vectorized<float>& a, c
   return at::vec::convert_from_float<scalar_t>(a, b);
 }
 
+// base type /* allow only f32, f16, bf16*/
+template <
+    typename scalar_t,
+    typename std::enable_if_t<is_reduced_floating_point_v<scalar_t> || std::is_same_v<scalar_t, float>, int> = 1>
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/vec.h]:
```suggestion
template <typename scalar_t,
                  typename std::enable_if_t<is_reduced_floating_point_v<scalar_t>, int> = 0>
```

File: sgl-kernel/csrc/cpu/vec.h [hunk 4]
```diff
@@ -16,6 +16,28 @@ inline Vectorized<scalar_t> convert_from_float_ext(const Vectorized<float>& a, c
...
+  using fVec = at::vec::Vectorized<float>;
+  bVec x_vec = bVec::loadu(data);
+  fVec x0, x1;
+  std::tie(x0, x1) = at::vec::convert_to_float(x_vec);
+  return std::make_tuple(x0, x1);
+}
+
+// specification for `float`
+template <>
+inline std::tuple<Vectorized<float>, Vectorized<float>> load_float_vec2<float>(const float* __restrict__ data) {
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/vec.h]:
```suggestion
inline std::tuple<Vectorized<float>, Vectorized<float>> load_float_vec2(const float* __restrict__ data) {
```

File: sgl-kernel/csrc/cpu/vec.h [hunk 5]
```diff
@@ -16,6 +16,28 @@ inline Vectorized<scalar_t> convert_from_float_ext(const Vectorized<float>& a, c
...
+inline std::tuple<Vectorized<float>, Vectorized<float>> load_float_vec2(const scalar_t* __restrict__ data) {
+  using bVec = at::vec::Vectorized<scalar_t>;
+  using fVec = at::vec::Vectorized<float>;
+  bVec x_vec = bVec::loadu(data);
+  fVec x0, x1;
+  std::tie(x0, x1) = at::vec::convert_to_float(x_vec);
+  return std::make_tuple(x0, x1);
+}
+
+// specification for `float`
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/vec.h]:
```suggestion
```

File: sgl-kernel/csrc/cpu/vec.h (hunk 4 above)
Comment (mingfeima) [reply sgl-kernel/csrc/cpu/vec.h]: let's see if using `overload` can be compiled. we can save 2 or 3 lines here.

File: sgl-kernel/csrc/cpu/common.h [hunk 6]
```diff
@@ -47,6 +47,47 @@ namespace {
     }                                                            \
   }()
 
+// dispatch mixed precision of TYPE2 with TYPE1:
+// TYPE2: float32
+//    with TYPE1: bfloat16, float16
+// TYPE2: bfloat16, float16
+//    with TYPE1: same as TYPE2
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/common.h]:
```suggestion
// dispatch with mixed dtypes (TYPE1, TYPE2):
//   TYPE1: the primary dtype (input, output, weight);
//   TYPE2: the secondary dtype (bias, etc.).
```

File: sgl-kernel/csrc/cpu/common.h [hunk 7]
```diff
@@ -47,6 +47,47 @@ namespace {
...
+          using scalar_t = at::Half;                               \
+          using param_t = at::Half;                                \
+          return __VA_ARGS__();                                    \
+        }                                                          \
+        default:                                                   \
+          TORCH_CHECK(false, "Unsupported floating data type.\n"); \
+      }                                                            \
+    }                                                              \
+  }()
+
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/common.h:88]: good job! exactly what i meant :)

**Review by mingfeima — APPROVED** [review]
generally LGTM! just a few minor changes needed!

File: sgl-kernel/csrc/cpu/topk.cpp (hunk 2 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/topk.cpp]: Yes, got the point, and refiend.

File: sgl-kernel/csrc/cpu/vec.h (hunk 4 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/vec.h]: yes, changed to use `overload`,

File: sgl-kernel/csrc/cpu/vec.h (hunk 5 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/vec.h]: removed

File: sgl-kernel/csrc/cpu/common.h (hunk 6 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/common.h]: Thanks, refined.

**Review by mingfeima — APPROVED** [review]

**Comment by jianan-gu** [comment]
Checked CI failures are not related to this PR.

**Comment by jianan-gu** [comment]
Hi @zhyncs, could you kindly help review this PR? Thanks!

**Review by Alcanderian — APPROVED** [review]

**Review by Alcanderian — APPROVED** [review]
